Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean

```diff
@@ -0,0 +1,967 @@
+
+/-- The proposition that a point class on `W` is nonsingular. If `P` is a point representative,
+then `W.nonsingular_lift ⟦P⟧` is definitionally equivalent to `W.nonsingular P`. -/
+@[pp_dot]
+def nonsingular_lift (P : PointClass R) : Prop :=
+  P.lift W.nonsingular fun _ _ => propext ∘ W.nonsingular_of_equiv
+
+@[simp]
+lemma nonsingular_lift_eq (P : Fin 3 → R) : W.nonsingular_lift ⟦P⟧ = W.nonsingular P :=
+  rfl
+
+lemma nonsingular_lift_zero [Nontrivial R] : W.nonsingular_lift ⟦![0, 1, 0]⟧ :=
+  W.nonsingular_zero
+
+lemma nonsingular_lift_zero' [NoZeroDivisors R] {Y : R} (hy : Y ≠ 0) :
+    W.nonsingular_lift ⟦![0, Y, 0]⟧ :=
+  W.nonsingular_zero' hy
+
+lemma nonsingular_lift_some (X Y : R) :
+    W.nonsingular_lift ⟦![X, Y, 1]⟧ ↔ W.toAffine.nonsingular X Y :=
+  W.nonsingular_some X Y
+
+variable {F : Type u} [Field F] {W : Projective F}
+
+lemma equiv_of_Zeq0 {P Q : Fin 3 → F} (hP : W.nonsingular P) (hQ : W.nonsingular Q) (hPz : P z = 0)
+    (hQz : Q z = 0) : P ≈ Q := by
+  rw [fin3_def P, hPz] at hP ⊢
+  rw [fin3_def Q, hQz] at hQ ⊢
+  simp [nonsingular_iff, equation_iff] at hP hQ
+  simp [pow_eq_zero hP.left.symm, pow_eq_zero hQ.left.symm] at *
+  exact ⟨Units.mk0 (P y / Q y) <| div_ne_zero hP hQ, by simp [div_mul_cancel _ hQ]⟩
+
+lemma equiv_zero_of_Zeq0 {P : Fin 3 → F} (h : W.nonsingular P) (hPz : P z = 0) : P ≈ ![0, 1, 0] :=
+  equiv_of_Zeq0 h W.nonsingular_zero hPz rfl
+
+lemma equiv_some_of_Zne0 {P : Fin 3 → F} (hPz : P z ≠ 0) : P ≈ ![P x / P z, P y / P z, 1] :=
+  ⟨Units.mk0 _ hPz, by simp [← fin3_def P, mul_div_cancel' _ hPz]⟩
+
+lemma nonsingular_iff_affine_of_Zne0 {P : Fin 3 → F} (hPz : P z ≠ 0) :
+    W.nonsingular P ↔ W.toAffine.nonsingular (P x / P z) (P y / P z) :=
+  (W.nonsingular_of_equiv <| equiv_some_of_Zne0 hPz).trans <| W.nonsingular_some ..
+
+lemma nonsingular_of_affine_of_Zne0 {P : Fin 3 → F}
+    (h : W.toAffine.nonsingular (P x / P z) (P y / P z)) (hPz : P z ≠ 0) : W.nonsingular P :=
+  (nonsingular_iff_affine_of_Zne0 hPz).mpr h
+
+lemma nonsingular_affine_of_Zne0 {P : Fin 3 → F} (h : W.nonsingular P) (hPz : P z ≠ 0) :
+    W.toAffine.nonsingular (P x / P z) (P y / P z) :=
+  (nonsingular_iff_affine_of_Zne0 hPz).mp h
+
+end Equation
+
+section Polynomial
+
+/-! ### Group operation polynomials -/
+
+/-- The $Y$-coordinate of the negation of a point representative. -/
+@[pp_dot]
+def negY (P : Fin 3 → R) : R :=
+  -P y - W.a₁ * P x - W.a₃ * P z
+
+lemma negY_smul (P : Fin 3 → R) (u : Rˣ) : W.negY (u • P) = u * W.negY P := by
+  simp only [negY, smul_fin3_ext]
+  ring1
+
+/-- The truncated $X$-coordinate of the addition of two point representatives, where their
+$Z$-coordinates are non-zero and their $X$-coordinates divided by $Z$-coordinates are distinct.
+This is only an auxiliary definition needed to define `WeierstrassCurve.Projective.addY'_of_Xne`,
+and the full $X$-coordinate is defined in `WeierstrassCurve.Projective.addX_of_Xne`. -/
+@[pp_dot]
+def addX'_of_Xne (P Q : Fin 3 → R) : R :=
```

Comment:
`negAddXOfXNe`?
(I don't care much since I won't need the projective addition formulas for the reduction homomorphism. The projective addition formulas have the same number of cases as the affine formulas, so we can just transfer the affine addition formulas through the Equiv on Points, and the projective formulas do not look more computable than the affine ones (basically you need to check the same amount of conditions).)